rapid_response_xblock/static/js/src/rapid.js

```diff
@@ -43,6 +46,227 @@
+      var tickValues = [];
+      var lookup = {};
+      var maxCount = domain.domain()[1];
+      if (!isNaN(maxCount)) {
+        var numTicks = 6;
+        var tickIncrement = Math.ceil(maxCount / numTicks);
+        for (var tickCount = 0; tickCount < numTicks; ++tickCount) {
+          var tick = tickCount * tickIncrement;
+          if (!(tick in lookup)) {
+            lookup[tick] = true;
+            tickValues.push(tick);
+          }
+        }
+      }
+      return tickValues;
+    }
+
+    function wrapText(textSelector, barWidth, oldText) {
+      // see https://bl.ocks.org/mbostock/7555321 for inspiration
+      // SVG doesn't have a capability to wrap text except for foreignObject which is not supported in IE11.
+      // So we have to calculate it manually
+      textSelector.each(function() {
+        var root = d3.select(this);
+
+        var rootY = root.attr("y");
+        var rootDy = parseFloat(root.attr("dy"));
+
+        root.selectAll("tspan").remove();
+        var words = oldText.split(/\s+/);
+        var tspan = root.append("tspan").attr("x", 0).attr("y", rootY).attr("dy", rootDy + "em");
+
+        var currentLine = 0;
+        var lineHeight = 1.1;
+        words.forEach(function(word) {
+          if (!word) {
+            return;
+          }
+
+          var oldText = tspan.text();
+          tspan.text(oldText + " " + word);
+          if (tspan.node().getComputedTextLength() > barWidth) {
+            currentLine++;
+            tspan.text(oldText + " ");
+            tspan = root.append("tspan").attr("x", 0).attr("y", rootY).attr(
+              "dy", ((currentLine * lineHeight) + rootDy) + "em"
+            ).text(word);
+          }
+        });
+      });
+    }
+
+    function renderD3() {
+      // This function should be called whenever the response information changes.
+
+      var message = chart.select(".message");
+      var responses = state.responses;
+      if (responses.length === 0) {
+        message.text("No data available").classed("hidden", false);
+      } else {
+        message.classed("hidden", true);
+      }
+
+      // Compute responses into information suitable for a bar graph.
+      var histogram = makeHistogram(responses);
+      var histogramLookup = {};
+      histogram.forEach(function(response) {
+        histogramLookup[response.answer_id] = response;
+      });
+
+      // Add answer ids to the color domain if they don't already exist
+      histogram.forEach(function(response) {
+        var answerId = response.answer_id;
+        if (!_.includes(colorDomain, answerId)) {
+          colorDomain.push(answerId);
+        }
+      });
+
+      // Create x scale to map answer ids to bar x coordinate locations. Note that
+      // histogram was previously sorted in order of the lowercase answer id.
+      var x = d3.scaleBand().rangeRound([0, chartWidth]).padding(0.1).domain(
+        histogram.map(function(value) {
+          return value.answer_id;
+        })
```

Comment:
1. There's some inconsistency in the parameter name you use when iterating over `histogram` elements. Here it's `value`, above it's `response`, below (line 206) it's `item`.
2. It looks like at least 2 separate times you're looping over `histogram` to pull out the `answer_id` from each object in the collection (here, line 180). Seems like it would make sense to just do that once and assign the value to a variable.